Add some more tests to parser

dionyziz opened the pull request.

dionyziz committed:
Add some more tests to parser

dionyziz committed:
ast: __repr__ for for expressions

dionyziz committed:
parser: Test for loops and fix for loop bug where all loops were downto

dionyziz committed:
parser test: patterns

dionyziz committed:
Add live test script for mac

dionyziz committed:
AST: Add generic __repr__ for all AST nodes to ease debugging

dionyziz committed:
livetest now depends on ast changes also

dionyziz committed:
Parser test: fix whitespace

dionyziz committed:
Parser test: letdef test

dionyziz committed:
parser test: delete test

Renelvon commented on compiler/parse.py:
```diff
@@ -317,7 +317,7 @@ def p_dim_expr(self, p):
     def p_for_expr(self, p):
         """for_expr : FOR GENID EQ expr DOWNTO expr DO expr DONE
                     | FOR GENID EQ expr TO expr DO expr DONE"""
-        if p[5] == 'TO':
```
Consider splitting this rule into 2 grammar rules, so as not to depend on (otherwise useless) value of the`TO` token.

dionyziz committed:
AST: Add precondition assertions for type parameter

dionyziz committed:
Parse: rename some parse rules for clarity

dionyziz committed:
Parse: Split simple_expr rules into several

dionyziz committed:
Parse: Split for_expr rules into several

PLY doesn't allow deducing which grammar rule matched
and this often causes bugs.

We had a bug with `for` where all `for`s were `downto`
`for`s due to a string comparison which is not the responsibility
of the parser, but the lexer.

This fix makes it harder to introduce such bugs.

dionyziz committed:
Type test: optimize=0

dionyziz committed:
Test type: More verbose functional testing

dionyziz committed:
Makefile: Put unit before functional tests

dionyziz committed:
AST test: Regression test for __eq__ use of .type

__eq__ in AST Nodes assumes two objects of the same class
have the same attributes (no attributes are added in runtime).

This is not true with the current type table, because it modifies
ast.Constructor objects to add the `type` attribute. This regression
test ensures equality works between ast.Constructor objects that
are constructed manually and those that are passed through the type
table.

This is fixed in #25 and the regression test should pass.

This is related to the fact that semantic checks are currently done
before parsing is completed - this will change in future versions.

dionyziz committed:
Parser test: Optimizations

* Memoize parsers at class level
* Use setUpClass() for simple expressions used across tests

dionyziz committed:
Parser test: def_list test

dionyziz committed:
Parser test: Don't test implementation detail states

dionyziz committed:
Parser test: Add test for function def & improve param list test

dionyziz committed:
Parser test: Beautify tests for shorter lines

dionyziz committed:
Parser test: Test that each built-in type is parsed

dionyziz committed:
Parser test: Test clauses and patterns

dionyziz committed:
Parser test: Include `mod` operator in boolean expression tests

dionyziz committed:
Parser test: begin/end, func call, constr call, simple expr seq

dionyziz committed:
Parser test: More expression tests

* array_expr
* paren_expr
* conid_expr
* genid_expr
* constr_pipe_seq
* tdef & seq
* typedef & type_seq

Renelvon commented:
Multiple build failures.
Local reproduction: Yes (on Linux/Python3.3/PLY 3.4).

Minimal failing example:

```
$ make clean; make unittest
```

Possible cause: Unit testing requires lexing tables.

Passes:

```
$ make clean; make prepare; make unittest; make functionaltests
```

Proposed fix: Change Makefile:

```
test: clean prepare unittest functionaltests
```

Renelvon commented on compiler/ast.py:
```diff
@@ -37,6 +37,18 @@ def copy_pos(self, node):
+                safe_values.append(value)
+            else:
+                safe_values.append('(non-scalar of type %s)' % value.__class__.__name__)
+        pairs = ["%s = '%s'" % (attr, value) for (attr, value) in zip(attrs, safe_values)]
```
An iterator would suffice:

```
pairs = ( ... )
```

A suggestion: Use `list`s only where a mutable container is required. If the container is intended to be immutable, use `tuple`s. If the container is not meant to be reused, avoid the container's construction altogether and just return an iterator.

https://en.wikipedia.org/wiki/Deforestation_(computer_science)

Renelvon commented on Makefile:
```diff
@@ -12,10 +12,14 @@ all: clean prepare test
+unittest:
 	nosetests
 
+functionaltests: $(BINPATH)/ptest.sh
```
Shouldn't it be `unittests`, then?

Renelvon commented on Makefile:
```diff
@@ -12,10 +12,14 @@ all: clean prepare test
 
-test: clean prepare $(BINPATH)/ptest.sh
-	$(BINPATH)/ptest.sh
+test: clean unittest prepare functionaltests
```
It turns out this is not the correct order, and for good reason. For efficiency, the unit tests assume an optimized lexer (with cached tables), but this is only produced during `prepare`. Putting `prepare` before `unittest` should in principle work fine; the parser tables will be ignored (and not regenerated), since all unit tests use an un-optimized `parser`. On the other hand, using the same lexer across all tests should not be a problem.

Renelvon commented on compiler/ast.py:
```diff
@@ -239,14 +245,18 @@ def __init__(self, condition, body):
 class VariableDef(Def):
     def __init__(self, name, type=None):
+        if type is not None:
+            assert(isinstance(type, Type))
```
This assertion is a bit problematic. The spec requires that the `type` be an instance of `Ref`, therefore a stricter assertion would be

```
assert(isinstance(type, Ref))
```

Though this is is better, it is still not the appropriate test. Suppose we try to parse the following

```
let mutable a : ref int
```

According to the spec, the `type` of `a` is `ref ref int`. Now suppose the parser has a bug and does not pack `ref int` appropriately and sets `type` to `ast.Ref(ast.Int())`. Then all assertions will pass, even though the parser failed to function correctly.

The correct way to test this is via unit testing: If the user supplies a type `T` to the definition, then the parser should set the `type` attribute to `ast.Ref(T)`, and similarly for other nodes (like `FunctionDef`, `ArrayVariableDef` and `TDef`).

In addition, all these `type`s should be validated, but this should probably be left for the semantic analysis pass.

Renelvon commented on compiler/parse.py:
```diff
@@ -248,17 +248,21 @@ def p_simple_expr(self, p):
+        p[0] = p[1]
         _track(p)
 
+    def p_array_simple_expr(self, p):
```
Good work. We should do more of this.

Renelvon commented on tests/test_ast.py:
```diff
@@ -2,10 +2,48 @@
 
 class TestAST(unittest.TestCase):
+# TODO: remove boilerplate code after simple_api branch gets merged
+    def setUp(self):
```
Use `setUpClass`?

dionyziz committed:
Split unit tests into independent python instances to avoid PLY caching

dionyziz committed:
Makefile: Change prepare/unittest order, as parser test requires lex tables

dionyziz committed:
Makefile: Rename rules for consistency

dionyziz committed:
AST: Replace list with iterator

dionyziz committed:
AST: Remove unnecessary assertions that are unit tests

dionyziz committed:
AST test: Use classmethods for memoization

Renelvon commented on compiler/parse.py:
```diff
@@ -248,17 +248,21 @@ def p_simple_expr(self, p):
 
+    def p_array_simple_expr(self, p):
+        """array_simple_expr : GENID LBRACKET expr_comma_seq RBRACKET"""
+        p[0] = ast.ArrayExpression(p[1], p[3])
```
Add line tracking.

Renelvon commented on compiler/parse.py:
```diff
@@ -248,17 +248,21 @@ def p_simple_expr(self, p):
+
+    def p_paren_simple_expr(self, p):
+        """paren_simple_expr : LPAREN expr RPAREN"""
+        p[0] = p[2]
```
Add line tracking.

Renelvon commented on compiler/parse.py:
```diff
@@ -248,17 +248,21 @@ def p_simple_expr(self, p):
+
+    def p_bang_simple_expr(self, p):
+        """bang_simple_expr : BANG simple_expr"""
+        p[0] = ast.UnaryExpression(p[1], p[2])
```
Add line tracking.

Renelvon commented on tests/test_ast.py:
```diff
@@ -2,10 +2,49 @@
 
 
 class TestAST(unittest.TestCase):
+# TODO: remove boilerplate code after simple_api branch gets merged
```
Use the new API: drop `lexer` and change `optimize=0` to `optimize=False`.

Renelvon commented on tests/test_parser.py:
```diff
@@ -6,13 +6,18 @@
+        cls.yfunc = cls._parse("let y = 2", "letdef")
+
+    @classmethod
     def _parse(self, data, start='program'):
```
Use the new API: drop `lexer` and change `optimize=0` to `optimize=False`.

Renelvon commented:
Please merge `master`.

dionyziz committed:
Parser: Add line tracking to some new rules missing it

dionyziz committed:
Makefile: Clean cached parse tables after unit tests

dionyziz committed:
Makefile: make clean before functional tests

dionyziz committed:
Type test: optimizations

dionyziz committed:
Merge branch 'master' into parser_test2

Conflicts:
	tests/test_type.py

Parser API has changed. Resolve conflicts by using the new API
in new type tests.

dionyziz committed:
Makefile: Remove duplicate call to clean/prepare

dionyziz committed:
Type test: Clear error log when reusing parser

dionyziz committed:
Error: Docstring in logger mock clear method

dionyziz committed:
Test: Use new parser API

dionyziz committed:
pep8

dionyziz committed:
Remove unused imports

dionyziz committed:
AST: Beautify ast node __repr__  - indentation for nested nodes

dionyziz committed:
AST: Remove duplicate definition for Expression

dionyziz committed:
Parser: Split array var def rules and use name var instead of variable

Renelvon commented on tests/test_type.py:
```diff
@@ -62,10 +62,18 @@ def test_function_types(self):
         mock = error.LoggerMock()
-        parse.parse(typeDefListList, logger=mock)
-        return mock.success
+        cls.parser = parse.Parser(logger=mock, optimize=0)
```
`optimize=False`

dionyziz committed:
Type tests: style fix
